File: src/proj4/projections/bonne.ts
import { HALF_PI } from '../constants';
import { WGS84Projection } from '.';
import { adjustLat, adjustLon, hypot, pjEnfn, pjInvMlfn, pjMlfn } from '../common';

import type { En } from '../common';
import type { ProjectionTransformDefinition } from '.';
import type { VectorPoint } from 's2-tools/geometry';

const EPS10 = 1e-10;
const { abs, sin, cos, sqrt, atan2, tan } = Math;

/**
 * # Bonne (Werner lat_1=90) Projection
 *
 * **Classification**: Miscellaneous
 *
 * **Available forms**: Forward and inverse, spherical and ellipsoidal
 *
 * **Defined area**: Global
 *
 * **Alias**: `bonne`
 *
 * **Domain**: 2D
 *
 * **Input type**: Geodetic coordinates
 *
 * **Output type**: Projected coordinates
 *
 * ## Projection String
 * ```
 * +proj=bonne +lat_1=10
 * ```
 *
 * ## Required Parameters
 * - `lat1`: Latitude of first standard parallel
 *
 * ## Optional Parameters
 * - `lon0`: Longitude of origin
 * - `ellps`: Ellipsoid name
 * - `R`: Radius of sphere
 * - `x0`: False easting
 * - `y0`: False northing
 *
 * ![Bonne (Werner lat_1=90) Projection](https://github.com/OSGeo/PROJ/blob/38dd7c2446f3500a43f0257f5a4833d6aa5aab0b/docs/source/operations/projections/images/bonne.png?raw=true)
 */
export class BonneWernerProjection
  extends WGS84Projection
  implements ProjectionTransformDefinition
{
  name = 'Bonne (Werner lat_1=90)';
  names = [this.name, 'bonne'];
  // BonneWernerProjection specific variables
  phi1 = 0;
  en: En = [0, 0, 0, 0, 0];
  m1: number = 0;
  am1: number = 0;
  cphi1: number = 0;

  /** Preps an Albers Conic Equal Area projection */
  constructor() {
    super();
    let c;

    this.phi1 = this.lat1;
    if (abs(this.phi1) < EPS10) {
      throw new Error('Invalid latitude');
    }
    if (this.es) {
      this.en = pjEnfn(this.es);
      this.m1 = pjMlfn(this.phi1, (this.am1 = sin(this.phi1)), (c = cos(this.phi1)), this.en);
      this.am1 = c / (sqrt(1 - this.es * this.am1 * this.am1) * this.am1);
      this.inverse = this.eInv;
      this.forward = this.eFwd;
    } else {
      if (abs(this.phi1) + EPS10 >= HALF_PI) {
        this.cphi1 = 0;
      } else {
        this.cphi1 = 1 / tan(this.phi1);
      }
      this.inverse = this.sInv;
      this.forward = this.sFwd;
    }
  }

  /**
   * Bonne Werner Easting forward equations--mapping lon-lat to x-y
   * @param p - lon-lat WGS84 point
   * @returns - an Bonne Werner point
   */
  eFwd(p: VectorPoint): VectorPoint {
    const lam = adjustLon(p.x - (this.long0 || 0));
    const phi = p.y;
    let E, c;
    const rh = this.am1 + this.m1 - pjMlfn(phi, (E = sin(phi)), (c = cos(phi)), this.en);
    E = (c * lam) / (rh * sqrt(1 - this.es * E * E));
    p.x = rh * sin(E);
    p.y = this.am1 - rh * cos(E);

    p.x = this.a * p.x + (this.x0 || 0);
    p.y = this.a * p.y + (this.y0 || 0);
    return p;
  }

  /**
   * Bonne Werner Easting inverse equations--mapping x-y to lon-lat
   * @param p - Bonne Werner point
   * @returns - lon-lat WGS84 point
   */
  eInv(p: VectorPoint): VectorPoint {
    p.x = (p.x - (this.x0 || 0)) / this.a;
    p.y = (p.y - (this.y0 || 0)) / this.a;

    let s: number, lam: number;
    const rh = hypot(p.x, (p.y = this.am1 - p.y));
    const phi = pjInvMlfn(this.am1 + this.m1 - rh, this.es, this.en);
    if ((s = abs(phi)) < HALF_PI) {
      s = sin(phi);
      lam = (rh * atan2(p.x, p.y) * sqrt(1 - this.es * s * s)) / cos(phi);
    } else if (abs(s - HALF_PI) <= EPS10) {
      lam = 0;
    } else {
      throw new Error();
    }
    p.x = adjustLon(lam + (this.long0 || 0));
    p.y = adjustLat(phi);
    return p;
  }

  /**
   * Bonne Werner Southing forward equations--mapping lon-lat to x-y
   * @param p - lon-lat WGS84 point
   * @returns - an Bonne Werner point
   */
  sFwd(p: VectorPoint): VectorPoint {
    const lam = adjustLon(p.x - (this.long0 || 0));
    const phi = p.y;
    let E: number;
    const rh = this.cphi1 + this.phi1 - phi;
    if (abs(rh) > EPS10) {
      p.x = rh * sin((E = (lam * cos(phi)) / rh));
      p.y = this.cphi1 - rh * cos(E);
    } else {
      p.x = p.y = 0;
    }

    p.x = this.a * p.x + (this.x0 || 0);
    p.y = this.a * p.y + (this.y0 || 0);
    return p;
  }

  /**
   * Bonne Werner Southing inverse equations--mapping x-y to lon-lat
   * @param p - Bonne Werner point
   * @returns - lon-lat WGS84 point
   */
  sInv(p: VectorPoint): VectorPoint {
    p.x = (p.x - (this.x0 || 0)) / this.a;
    p.y = (p.y - (this.y0 || 0)) / this.a;

    let lam: number;
    const rh = hypot(p.x, (p.y = this.cphi1 - p.y));
    const phi = this.cphi1 + this.phi1 - rh;
    if (abs(phi) > HALF_PI) {
      throw new Error();
    }
    if (abs(abs(phi) - HALF_PI) <= EPS10) {
      lam = 0;
    } else {
      lam = (rh * atan2(p.x, p.y)) / cos(phi);
    }
    p.x = adjustLon(lam + (this.long0 || 0));
    p.y = adjustLat(phi);
    return p;
  }
}

File: rust/proj/project/bonne.rs
use crate::proj::{
    BONNE, CoordinateStep, EPS10, LATITUDE_OF_FIRST_STANDARD_PARALLEL, Proj, ProjMethod,
    ProjectCoordinates, TransformCoordinates, enfn, inv_mlfn, mlfn,
};
use alloc::vec::Vec;
use core::{cell::RefCell, f64::consts::FRAC_PI_2};
use libm::{atan2, copysign, cos, fabs, hypot, sin, sqrt, tan};

/// Bonne Variables
#[derive(Debug, Default, Clone, PartialEq)]
pub struct BonneData {
    phi1: f64,
    cphi1: f64,
    am1: f64,
    m1: f64,
    en: Vec<f64>,
}

/// # Bonne (Werner lat_1=90) Projection
///
/// **Classification**: Miscellaneous
///
/// **Available forms**: Forward and inverse, spherical and ellipsoidal
///
/// **Defined area**: Global
///
/// **Alias**: `bonne`
///
/// **Domain**: 2D
///
/// **Input type**: Geodetic coordinates
///
/// **Output type**: Projected coordinates
///
/// ## Projection String
/// ```ini
/// +proj=bonne +lat_1=10
/// ```
///
/// ## Required Parameters
/// - `lat1`: Latitude of first standard parallel
///
/// ## Optional Parameters
/// - `lon0`: Longitude of origin
/// - `ellps`: Ellipsoid name
/// - `R`: Radius of sphere
/// - `x0`: False easting
/// - `y0`: False northing
///
/// ![Bonne (Werner lat_1=90) Projection](https://github.com/Open-S2/gis-tools/blob/master/assets/proj4/projections/images/bonne.png?raw=true)
#[derive(Debug, Clone, PartialEq)]
pub struct BonneProjection {
    proj: RefCell<Proj>,
    store: RefCell<BonneData>,
    method: ProjMethod,
}
impl ProjectCoordinates for BonneProjection {
    fn code(&self) -> i64 {
        BONNE
    }
    fn name(&self) -> &'static str {
        "Bonne"
    }
    fn names() -> &'static [&'static str] {
        &["Bonne (Werner lat_1=90)", "bonne_werner", "Bonne", "bonne"]
    }
}
impl CoordinateStep for BonneProjection {
    fn new(proj: RefCell<Proj>) -> Self {
        let mut store = BonneData::default();
        let method: ProjMethod;
        {
            let proj = &mut proj.borrow_mut();
            store.phi1 = proj.params.get(&LATITUDE_OF_FIRST_STANDARD_PARALLEL).unwrap().f64();
            if fabs(store.phi1) < EPS10 {
                panic!("Invalid value for lat_1: |lat_1| should be > 0");
            }

            method = if proj.es != 0.0 {
                store.en = enfn(proj.n);
                store.am1 = sin(store.phi1);
                let c = cos(store.phi1);
                store.m1 = mlfn(store.phi1, store.am1, c, &store.en);
                store.am1 = c / (sqrt(1. - proj.es * store.am1 * store.am1) * store.am1);
                ProjMethod::Ellipsoidal
            } else {
                if fabs(store.phi1) + EPS10 >= FRAC_PI_2 {
                    store.cphi1 = 0.;
                } else {
                    store.cphi1 = 1. / tan(store.phi1);
                }
                ProjMethod::Spheroidal
            };
        }
        BonneProjection { proj, store: store.into(), method }
    }
    fn forward<P: TransformCoordinates>(&self, p: &mut P) {
        if self.method == ProjMethod::Ellipsoidal {
            bonne_e_forward(&mut self.store.borrow_mut(), &self.proj.borrow(), p);
        } else {
            bonne_s_forward(&mut self.store.borrow_mut(), &self.proj.borrow(), p);
        }
    }
    fn inverse<P: TransformCoordinates>(&self, p: &mut P) {
        if self.method == ProjMethod::Ellipsoidal {
            bonne_e_inverse(&mut self.store.borrow_mut(), &self.proj.borrow(), p);
        } else {
            bonne_s_inverse(&mut self.store.borrow_mut(), &self.proj.borrow(), p);
        }
    }
}

/// Bonne Ellipsoidal Forward
pub fn bonne_e_forward<P: TransformCoordinates>(bonne: &mut BonneData, proj: &Proj, p: &mut P) {
    let mut e = sin(p.phi());
    let c = cos(p.phi());
    let rh = bonne.am1 + bonne.m1 - mlfn(p.phi(), e, c, &bonne.en);
    if fabs(rh) > EPS10 {
        e = c * p.lam() / (rh * sqrt(1. - proj.es * e * e));
        p.set_x(rh * sin(e));
        p.set_y(bonne.am1 - rh * cos(e));
    } else {
        p.set_x(0.);
        p.set_y(0.);
    }
}

/// Bonne Spheroidal Forward
pub fn bonne_s_forward<P: TransformCoordinates>(bonne: &mut BonneData, _proj: &Proj, p: &mut P) {
    let rh = bonne.cphi1 + bonne.phi1 - p.phi();
    if fabs(rh) > EPS10 {
        let e = p.lam() * cos(p.phi()) / rh;
        p.set_x(rh * sin(e));
        p.set_y(bonne.cphi1 - rh * cos(e));
    } else {
        p.set_x(0.);
        p.set_y(0.);
    }
}

/// Bonne Spheroidal Inverse
pub fn bonne_s_inverse<P: TransformCoordinates>(bonne: &mut BonneData, _proj: &Proj, p: &mut P) {
    p.set_y(bonne.cphi1 - p.y());
    let rh = copysign(hypot(p.x(), p.y()), bonne.phi1);
    let phi = bonne.cphi1 + bonne.phi1 - rh;
    let lam: f64;
    let abs_phi = fabs(phi);
    if abs_phi > FRAC_PI_2 {
        panic!("Coordinate outside projection domain");
    }
    if FRAC_PI_2 - abs_phi <= EPS10 {
        lam = 0.;
    } else {
        let lm = rh / cos(phi);
        if bonne.phi1 > 0. {
            lam = lm * atan2(p.x(), p.y());
        } else {
            lam = lm * atan2(-p.x(), -p.y());
        }
    }

    p.set_phi(phi);
    p.set_lam(lam);
}

/// Bonne Ellipsoidal Inverse
pub fn bonne_e_inverse<P: TransformCoordinates>(bonne: &mut BonneData, proj: &Proj, p: &mut P) {
    p.set_y(bonne.am1 - p.y());
    let rh = copysign(hypot(p.x(), p.y()), bonne.phi1);
    let phi = inv_mlfn(bonne.am1 + bonne.m1 - rh, &bonne.en);
    let lam: f64;
    let abs_phi = fabs(phi);
    if abs_phi < FRAC_PI_2 {
        let sinphi = sin(phi);
        let lm = rh * sqrt(1. - proj.es * sinphi * sinphi) / cos(phi);
        if bonne.phi1 > 0. {
            lam = lm * atan2(p.x(), p.y());
        } else {
            lam = lm * atan2(-p.x(), -p.y());
        }
    } else if abs_phi - FRAC_PI_2 <= EPS10 {
        lam = 0.;
    } else {
        panic!("Coordinates are outside the projection domain");
    }
    p.set_phi(phi);
    p.set_lam(lam);
}
